[MRG+1] ENH: Adds FunctionTransformer

llllllllll opened the pull request:
CallableTransformer allows a user to convert a standard python callable
into a transformer for use in a Pipeline.

Addresses: https://github.com/scikit-learn/scikit-learn/issues/3560

Commit:
ENH: Adds CallableTransformer

CallableTransformer allows a user to convert a standard python callable
into a transformer for use in a Pipeline.

amueller commented on doc/modules/preprocessing.rst:
```diff
@@ -423,3 +423,24 @@ values than observed values.
+Custom Transformers
+===================
+
+Often, you will want to convert an existing python function into transformer to
```
"into *a* transformer"

amueller commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+        func. If validate is false, there will be no input validation.
+    accept_sparse : boolean, optional
+        Indicate that func accepts a sparse array as input.
+    args: tuple, optional
```
missing space

amueller commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+        Indicate that the input X array should be checked before calling
+        func. If validate is false, there will be no input validation.
+    accept_sparse : boolean, optional
+        Indicate that func accepts a sparse array as input.
```
sparse matrix

amueller commented:
lgtm

llllllllll commented:
@GaelVaroquaux Could you take a look please?

mblondel commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+        be passed after X and y.
+    kwargs : dict, optional
+        A dictionary of keyword arguments to be passed to func.
+
```
Could you add a simple / short example here?

mblondel replied:
The same as in the user guide is fine.

ogrisel replied:
+1 for including a simple example as a doctest in the docstring of the class.

vene replied:
Suggestion for example: `func=partial(getattr, 'data')`, and feed the transformer a dict `{'data': X, other stuff...}`
Sorry, I can't think of better names.

I often end up doing this when I do e.g. text classification on conversational data and I have messages in both directions. I store my samples as {'from': from, 'to': to} and use a `FeatureUnion` of two pipelines, each grabbing the respective field and then doing a CountVectorizer.

jnothman replied:
except that `partial(getattr, 'data')` is more-or-less `operator.attrgetter`

GaelVaroquaux commented:
Sorry, @llllllllll, I am too tired to review this tonight. The jetlag is still killing me.

llllllllll commented:
@GaelVaroquaux No worries

ogrisel commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+    func : callable, optional default=None
+        The callable to use for the transformation. This will be passed
+        the same arguments as transform, with args and kwargs forwarded.
+        If func is None, then func will be the identity function.
```
Please insert one blankline before the documentation of the next parameter.

arjoly commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+        A tuple of positional arguments to be passed to func. These will
+        be passed after X and y.
+    kwargs : dict, optional
+        A dictionary of keyword arguments to be passed to func.
```
Do we need those? Can we use instead [partial](https://docs.python.org/3.4/library/functools.html)?

vene replied:
+1 for suggesting that users use `partial` (and exemplifying it somewhere). Implementing it just adds another hole where a bug could crawl.

arjoly commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+        Indicate that the input X array should be checked before calling
+        func. If validate is false, there will be no input validation.
+    accept_sparse : boolean, optional
+        Indicate that func accepts a sparse matrix as input.
```
Should it be made by  the function `self.func`?

vene commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class CallableTransformer(BaseEstimator, TransformerMixin):
+    """Allows the construction of a transformer from an arbitrary callable.
```
I suggest a more direct and brief "Constructs a transformer from an arbitrary callable."

vene commented on sklearn/preprocessing/callable_transformer.py:
```diff
@@ -0,0 +1,44 @@
+    def transform(self, X, y=None):
+        if self.validate:
+            X = check_array(X, self.accept_sparse)
+        return (self.func or (lambda X, y, *args, **kwargs: X))(
```
I'd explicitly do (omitting args)

```
if self.func is None:
    func = lambda X, y: X
else:
    func = self.func

return func(X, y)
```
BTW, would this class pickle well with the identity lambda? :confused:

amueller replied:
The pickle tests should be run on all transformers.

vene commented on sklearn/utils/estimator_checks.py:
```diff
@@ -131,7 +131,8 @@ def _yield_transformer_checks(name, Transformer):
         yield check_transformer_data_not_an_array
     # these don't actually fit the data, so don't raise errors
-    if name not in ['AdditiveChi2Sampler', 'Binarizer', 'Normalizer']:
+    if name not in ['AdditiveChi2Sampler', 'Binarizer',
```
Very silly comment, but these seemed alphabetically ordered, so it can't hurt to keep the order.

amueller replied:
Actually I think it can be removed here now, and probably AdditiveChi2Sampler can be removed, too, iirc

llllllllll replied:
sklearn.tests.test_common.test_non_meta_estimators

Removing these causes errors in `sklearn.tests.test_common.test_non_meta_estimators`

amueller replied:
what check fails?

llllllllll replied:
```
======================================================================
FAIL: sklearn.tests.test_common.test_non_meta_estimators('AdditiveChi2Sampler', <class 'sklearn.kernel_approximation.AdditiveChi2Sampler'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/joejev/.virtualenvs/qexec/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/joejev/projects/python/scikit-learn/sklearn/utils/estimator_checks.py", line 328, in check_transformer_general
    _check_transformer(name, Transformer, X, y)
  File "/home/joejev/projects/python/scikit-learn/sklearn/utils/estimator_checks.py", line 415, in _check_transformer
    assert_raises(ValueError, transformer.transform, X.T)
AssertionError: ValueError not raised


======================================================================
FAIL: sklearn.tests.test_common.test_non_meta_estimators('FunctionTransformer', <class 'sklearn.preprocessing.function_transformer.FunctionTransformer'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/joejev/.virtualenvs/qexec/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/joejev/projects/python/scikit-learn/sklearn/utils/estimator_checks.py", line 328, in check_transformer_general
    _check_transformer(name, Transformer, X, y)
  File "/home/joejev/projects/python/scikit-learn/sklearn/utils/estimator_checks.py", line 415, in _check_transformer
    assert_raises(ValueError, transformer.transform, X.T)
AssertionError: ValueError not raised


======================================================================
FAIL: sklearn.tests.test_common.test_non_meta_estimators('FunctionTransformer', <class 'sklearn.preprocessing.function_transformer.FunctionTransformer'>)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/joejev/.virtualenvs/qexec/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/joejev/projects/python/scikit-learn/sklearn/utils/estimator_checks.py", line 350, in check_transformers_unfitted
    assert_raises((AttributeError, ValueError), transformer.transform, X)
AssertionError: (<type 'exceptions.AttributeError'>, <type 'exceptions.ValueError'>) not raised
```

amueller replied:
ah ok, then never mind, leave it as is.

vene commented on sklearn/preprocessing/tests/test_callable_transformer.py:
```diff
@@ -0,0 +1,79 @@
+    return _func
+
+
+def test_delegate_to_func():
```
Would these tests simplify if dropping support for `args` and `kwargs` in favor of `partial`?

vene commented on doc/modules/preprocessing.rst:
```diff
@@ -423,3 +423,24 @@ values than observed values.
+===================
+
+Often, you will want to convert an existing python function into a transformer
+to assist in data cleaning or processing. Users may implement a transformer from
```
Reading this I might be confused. "Why will I want this? I can just do X = np.log(X) or something".
I think the most valuable use case is to grid search over the transformer function used. The paragraph might read better with a mention of this scenario.

mblondel replied:
Also, most people don't know what a callable is.

amueller replied:
@mblondel would you still stick with the name then?
I agree on @vene's comment.

jnothman replied:
I don't agree that it's most valuable with a search over the function. But I think it has no value except in a `Pipeline` or `FeatureUnion` and it should be illustrated in that context.

amueller replied:
Whoops that is what I meant. It is useful for pipelines. I don't think searching over the function is the main use case. I think selecting columns / slicing etc is a good use case.

vene commented on examples/preprocessing/plot_callable_transformer.py:
```diff
@@ -0,0 +1,69 @@
+    return X[:, 1:]
+
+
+def drop_first_component(X, y):
```
I'm not convinced that factoring this out as a function is the best example to set, since it is just standard pipeline use combined with a train_test_split. I would just do everything in the `__main__`.

llllllllll replied:
This just factors the work being done in to named components so that users can read through the code more easily. I don't mind changing it but I am not convinced that a single linear script is more readable.

vene replied:
well, I'd argue that the function named `drop_first_component` does more than its title may suggest. It just strikes me as an odd factorization.

vene commented on examples/preprocessing/plot_callable_transformer.py:
```diff
@@ -0,0 +1,69 @@
+    )), np.hstack((np.zeros(1000), np.ones(1000)))
+
+
+def all_but_first_column(X, y):
```
This shouldn't take a `y`, and I think the code shouldn't break if `func` only takes X, but `y` exists and is passed.

vene commented:
I'm concerned about what happens when a `y` is passed, but the `func` doesn't take/use it.

We fundamentally want this to work with `np.log`, `np.sqrt`, etc. But it looks to me like with the current implementation, we can't do `CallableTransformer(np.log).transform(X, y)`. Wouldn't this impact the use in supervised pipelines?

llllllllll commented:
This is a good point, do you think that this should accept an optional function to act on y?

vene commented:
I think it needs to be the same function, as it might need to use the value of `y` in deciding what to do to each row of `X` (e.g. class weights)

The very simplest solution is to just not allow this, and just ignore the `y`.
 
The next simplest solution is probably to add an attribute (`uses_y=True|False`), but that would make it cumbersome to grid search over callables where some use y and others don't.

I don't know what other solutions are, introspecting the callable, catching the exception? In which case maybe the call should be `func(X, y=y)`? This can get messy.

vene commented:
If we agree that a supervised pipeline with `func=np.log` should work by default, maybe we should have a test for this use case?

llllllllll commented:
Maybe we can have it accept a tuple for the `func` slot, where the tuple reads: `(func, pass_y)` and if it is a non-tuple, just treat it like: `(func, False)`

llllllllll commented:
After thinking about this a bit more, I think that a good thing might just be to make the call:

```python
try:
    return func(X, y)
except TypeError:
    return func(X)
```

I _would_ say that using a `pass_y` flag would be better; however, the search use case seems to make that not work.

Also, to address the `partial` vs `args, kwargs`, this was chosen to make searching params to the func easier.

vene commented:
I understand your argument. But this also adds more of a maintenance burden.

Also, it might not be that simple to do `try: return func(X, y)`. `np.log` has a call signature `(x[, out])` so it could actually succeed, or fail with a different exception than expected.

Simply not returning `TypeError` when called with two arguments cannot be unambiguously used to signal the correct signature here. Something like `func(X, y=y)` might be a bit better, but still, when the user runs into that one 3rd party library function that has an attribute named `y` by coincidence, she will have to write a clumsy lambda wrapper.

What's the most common use case here? I think it's when `y` isn't used at all. I also guess that grid searching over `func`s that use `y` and that don't use `y` isn't that common. I'm afraid we're rushing to generalize to scenarios that nobody actually needs.

So my preferences would be, in order (and with a pretty big gap between 2 and 3):
  1. simply don't support using `y` at all, just use `X`
  2. have a parameter indicating whether `y` is expected or not, and fail accordingly if it's not passed
  3. the (func, flag) tuple input
  4. do heuristics to detect whether `y` should be used at all.

As for kwargs vs partial, I still vote for partial. Grid search would still look alright with:
```
grid = dict(functransformer__func=(
    partial(select_columns, cols)
    for cols in [
        (0, 1),  # sepal width & petal width
        (2, 3)  # sepal length & petal length
    ]))
```

And we wouldn't have to reimplement stdlib functionality.

jnothman commented:
* I prefer having a parameter for "use y". I think the case of a search that mixes callables that require or do not require `y` means the user needs to compensate by making all their callables behave the same way, etc.
* seeing as this is one area users are likely to use anonymous functions, their incompatibility with pickling should probably be noted in the documentation.
* partial is better than kwargs as a constructor parameter. It ties the kwargs to the callable itself, while modifying the parameters through a kwargs param is just as cumbersome as modifying the callable itself; apart from which, I think the case where the kwargs are to be varied in a search is necessarily beyond the scope of a quick helper like this.
* perhaps the estimator should allow `inverse_transform` to also be provided, or we could decide that that is beyond its purpose.
* I think `FunctionTransformer` might be more understandable to a breadth of users.
* I'm not sure if this belongs in preprocessing; I think it is closely thematically coupled with `Pipeline` as a model composition utility.

amueller commented:
+1 on what @jnothman said. inverse_transform would be great, and an argument to pass y would be, too.

I am not entirely certain about preprocessing vs pipeline module. This only really makes sense when using a pipeline, which I think is a good argument.

amueller commented:
@llllllllll Sorry if this is taking more of your time than you anticipated.

glouppe commented:
Jumping into the conversation, as I have written myself several times such as class for personal use. However, I often find myself needing to apply a function element-wise rather than on the full X. In this setting, I dont know if proposing a shortcut for vectorizing a user-defined function would be something to consider? (using numpy.vectorize internally, this is easy) I fear this kind of use case might pop up sooner than later once such a transformer would be shipped in scikit-learn. 

Just my 2 cents -- I dont want this make this longer than it should :)

jnothman commented:
Can you give examples of element-wise functions that aren't compositions of
numpy functions?

On 4 June 2015 at 05:12, Gilles Louppe <notifications@github.com> wrote:

> Jumping into the conversation, as I have written myself several times such
> as class for personal use. However, I often find myself needing to apply a
> function element-wise rather than on the full X. In this setting, I dont
> know if proposing a shortcut for vectorizing a user-defined function would
> be something to consider? (using numpy.vectorize internally, this is easy)
> I fear this kind of use case might pop up sooner than later once such a
> transformer would be shipped in scikit-learn.
>
> Just my 2 cents -- I dont want this make this longer than it should :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4798#issuecomment-108583966>
> .
>

mblondel commented:
@amueller I prefer `FunctionTransformer` too.

llllllllll commented:
Just so everyone knows, I have not forgotten about this PR; however, I have been busy with work. I will address the comments made sometime this weekend so that they can go under another round of review.

amueller commented:
Thanks @llllllllll, your contribution is much appreciated :)

Commit:
ENH: Renames CallableTransformer -> FunctionTransformer.

Makes `pass_y` an argument to FunctionTransformer to indicate that the
labels should be passed to the wrapped function.

Commit:
COMPAT: Makes test_function_transformer py2 compatible.

jnothman commented on sklearn/preprocessing/function_transformer.py:
```diff
@@ -0,0 +1,52 @@
+
+    validate : bool, optional default=True
+        Indicate that the input X array should be checked before calling
+        func. If validate is false, there will be no input validation.
```
Note that this will ensure the input is a non-empty, 2-dimensional array (or sparse matrix) of finite numbers.

larsmans replied:
@jnothman Solved in ec7ddcb50e98323224d89255155068c65887b614.

larsmans committed:
DOC expand FunctionTransform docstring

llllllllll committed:
Merge pull request #1 from larsmans/functransf

DOC expand FunctionTransform docstring

larsmans commented:
I think we can keep inverses out of this PR. I'm ok with this living in `sklearn.pipeline` or `preprocessing`.

Merge conflict, please rebase or merge in master.

amueller commented:
Merged via #5059. Thanks everybody, in particular @llllllllll for his contribution :)

larsmans commented:
As Olivier would say, :beers:

jnothman commented:
HurraH!

On 4 August 2015 at 06:13, Lars <notifications@github.com> wrote:

> As Olivier would say, [image: :beers:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4798#issuecomment-127393859>
> .
>
